### PR TITLE
Refactoring in ilCtrlStructureReader

### PR DESCRIPTION
**Pull request by klees**
Hi @alex40724,

I did some refactoring in the `ilCtrlStructureReader` in an attempt to prepare it to be moved to an artifact. I think we are not quite there yet, but I also think this is a good point to show my work and (hopefully) get this merged.

What I did so far:
* I restructured the parsing process in `ilCtrlStructureReader` and broke it apart in several pieces. I wrote unit tests in that process where possible.
* I introduced `ilCtrlStructure` to hold the actual control structure instead of using plain arrays and wrote unit tests for this too.

The end result might be looking fairly different then the state from where I started, but I think it should be possible to follow my moves via the various commits and see that they are quite digestible 
when viewed individually.

There still are some more things I would do before introducing the `ilCtrlStructure` as artifact (roughly):
* Move creation of node-ids out of the database.
* Introduce a repository interface for `ilCtrlStructure` and implement it for the DB-version.
* Enhance `ilCtrlStructure` and use it together with the repository class in `ilCtrl`.
* Implement the repository for the artifact case.

I'm looking forward to your feedback!

Best regards!

### CODE REVIEW

**Comment by Amstutz** [comment]
Hi @klees

Note I am not maintainer for any of the involved components. This feedback is therefore done purely for the interest and the aim of helping improving this crucial component. Take or leave any parts of it.

First of all thx a lot for this step and the involved investigation. Special thx for adding the `$this->executed` var and for adding tests.

Question:
- Return Type Internal Classes: You do not use Return types in internal classes, why? (see: https://github.com/ILIAS-eLearning/ILIAS/pull/2525/files#diff-68092f71c075267b8412c87df7030ac9R11) 
- Heredoc Syntax: Your usage of the HEREDOC syntax in tests looks funky and makes it somewhat hard to read. Why not just put those things in some external php file as strings, and load it from there? IMO this could make that tests easier to read. However this is most probably a matter of opinion.
- Output Arguments: In several places you use ilCtrlStructure as output argument (argument which has the purpose of being changed in the function and then handed back, e.g.: https://github.com/ILIAS-eLearning/ILIAS/pull/2525/files#diff-ec75fce4e648fd5ebcf010ef38186709R97). I do not like this pattern. It is hard to track what is done with ilCtrlStructure. One has to jump very far down in the level of detail. Also note that you pass the output argument in first in parseFileTo and last in readDirTo. Could those functions not be slightly renamed and hand back a list or similar?

Change the following:
- Return Type: Add bool return type here: https://github.com/ILIAS-eLearning/ILIAS/pull/2525/files#diff-ec75fce4e648fd5ebcf010ef38186709R156

Thx a lot for this nice contribution.

Kindly
@amstutz

**Comment by klees** [comment]
Hi @Amstutz,

thanks for the feedback!

> Return Type: Add bool return type here

Good catch. [Changed and found another one!](https://github.com/ILIAS-eLearning/ILIAS/pull/2525/commits/9aac1624c918915be2c4e91cc941057c9d220997).

> Return Type Internal Classes: You do not use Return types in internal classes, why?

Honestly: for no good reason. These classes to me feel like throw away code that basically does nothing besides giving me access to some internal methods for testing, so I don't pay much attention here. It's inconsistent, though, because I did hint on the arguments...

> Heredoc Syntax: [..] Why not just put those things in some external php file as strings, and load it from there?

Yes, this looks funky. I do not put in into external files because the assertion that I write down below is intimately tangled with that very string and I don't want them to be far from each other. But opinion noted... Would be interesting to hear other voices here.

> Output Arguments

I struggle to understand this one. You don't like the pattern where I pass some structure into a function/method and then get back an updated version of that structure? Why? To me this is the very essence of why immutable structures are so powerful: we can clearly see the flow of information here.

In this particular case (`readDirTo` and `parseFileTo`) I definitely see why this might be confusing. For one, I chose different orders of parameters, which could definitely be improved. On the other hand, I only have that `ilCtrlStructure` argument to `readDirTo` to accomodate the test cases `ilCtrlStructureReaderTest::testReadRemovesDuplicateCallsInDatabase` and `ilCtrlStructureReaderTest::testReadRemovesDuplicateFilesInDatabaseIfCompPrefixIsSet`. Which I basically need because currently the database is tangled with the parsing process. Finally, I started this refactoring in a state, where the ctrl structure was updated as a side effect on some member variables in the parsing process. This side effect became explicit in the argument and return type. So basically, no new flow of information here, just transparency, forced by the immutable structure.

I don't understand your question regarding the names or returning a list. Could you elaborate on that? Also I'm very much interested to hear your opinion regarding the pattern.

Also, this might just be some issue to look into when refactoring continues. I think it might vanish when some more code is pulled out of the `ilCtrlStructureReader`.

Best regards!

**Review by alex40724 — APPROVED** [review]
These changes all look good to me. You may merge this, if the tests succeed.

Thanks!

Heredoc is ok to me, it is similar to the KS tests (there are also some very long strings), and moving the strings to files seems to decrease readability at least to me. (I have to open and switch to another file in my IDE).

Using with/immutable ilCtrlStructure objects and passing/returning them is also ok for me.

Introducing repo classes also seems to be a feasible next step.

I wonder what your plans are with the node-ids. It is a common issue that they change with updates. Would you like to change the ID generation? This could also be a first step towards saner URL routes in ILIAS.

**Comment by Amstutz** [comment]
Hi

Thanks for the feedback. As said, I am clompletely fine with the changes.

> Heredoc is ok to me, it is similar to the KS tests (there are also some very long strings), and moving the strings to files seems to decrease readability at least to me. 

Oke, thx for the feedback. This was part of why I was asking. I do not like our rendering tests in the UI Framework. They are hard to maintain, we need to improve that at some point. Obviously for other reasons than HEREDOC. BUT one of it is that I suspect long inline string to be a source of issue. If moving to files is helpful, I do not know. Maybe not. Thx for your opinion here.

> I struggle to understand this one. You don't like the pattern where I pass some structure into a function/method and then get back an updated version of that structure? Why?

I am completely fine with your way of doing it. However here my reasoning behind my question:

Imo it is less explizit to say, 
`foo = CreateANewStructureFromFooAndPutEverythingOfInterestOfBarIntoIt(foo,bar);`

Why would you want this additional complexity of bar in this function/method? Does it not make it harder to test (all those special cases with non empty bars, empty bars, bars with coliding values etc.)? Why enforce the need to generate to pass `new Bar();` in case Bars does not yet exist?

Than: 
`foo = generateFooFromEverythingOfInterestOfBar(bar,[maybePassFooFactory]).mergeWith(foo);`

This way the function/method is kept clean of any deps to non-empty bars/collisions etc. All that is moved to mergeWith if it is needed at all. Further the logic needed to handle all those cases, are already factored out in case anywhere else it could be handy to merge to foo's. However, it is slightly longer.

> I don't understand your question regarding the names or returning a list. Could you elaborate on that? Also I'm very much interested to hear your opinion regarding the pattern.

Forget about it. One could also pass back raw values as so:

foo.addValues(getEverythingOfInterestOfBarAsRawValuesList());

**Comment by klees** [comment]
Hi @alex40724!

Thanks for your feedback! Tests pass now.

> I wonder what your plans are with the node-ids. It is a common issue that they change with updates. Would you like to change the ID generation? This could also be a first step towards saner URL routes in ILIAS.

This definitely sounds interesting, but I wonder how this could work. I mean, basically we could just use class names instead of ids, but this would imply longer URLs and maybe even have security implications, and I don't think we want this. On the other hand I can't see how we could shorten the class names substantially, with a stable result and no collisions. Currently I would not touch this. It think it would be worthwhile to move it out of the database, though.

Best regards!

**Comment by klees** [comment]
Hi @Amstutz!

Thanks for you answer, you definitely have a point there.

I think you are suggesting to use something called a monoid and move the merge operation there. A monoid is a structure that supports the notion of some "null element" or "empty element" together with an associative "add" or "append" operation on elements. Like integers or lists. The parsing of every file would then create instances of that monoid that later are merged together via add. I indeed would expect that this leads to a cleaner and simpler structure. Monoids often are a nice ingredient to (de-)composable solutions.

I didn't notice this until now and it probably would be a good further improvement.

Thx!

**Pull request merged**
